boskos/ranch/ranch.go

```diff
@@ -0,0 +1,305 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package ranch
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"sync"
+	"time"
+
+	"github.com/Sirupsen/logrus"
+	"k8s.io/test-infra/boskos/common"
+)
+
+// Ranch is the place which all of the Resource objects lives.
+type Ranch struct {
+	Resources   []common.Resource
+	lock        sync.RWMutex
+	storagePath string
+}
+
+// Public errors:
+
+// OwnerNotMatch will be returned if request owner does not match current owner for target resource.
+type OwnerNotMatch struct {
+	request string
+	owner   string
+}
+
+func (o OwnerNotMatch) Error() string {
+	return fmt.Sprintf("OwnerNotMatch - request by %s, currently owned by %s", o.request, o.owner)
+}
+
+// ResourceNotFound will be returned if requested resource does not exist.
+type ResourceNotFound struct {
+	name string
+}
+
+func (r ResourceNotFound) Error() string {
+	return fmt.Sprintf("Resource %s not exist", r.name)
+}
+
+// StateNotMatch will be returned if requested state does not match current state for target resource.
+type StateNotMatch struct {
+	expect  string
+	current string
+}
+
+func (s StateNotMatch) Error() string {
+	return fmt.Sprintf("StateNotMatch - expect %v, current %v", s.expect, s.current)
+}
+
+// NewRanch creates a new Ranch object.
+// In: config - path to resource file
+//     storage - path to where to save/restore the state data
+// Out: A Ranch object, loaded from config/storage, or error
+func NewRanch(config string, storage string) (*Ranch, error) {
+
+	newRanch := &Ranch{
+		storagePath: storage,
+	}
+
+	if storage != "" {
+		buf, err := ioutil.ReadFile(storage)
+		if err == nil {
+			logrus.Infof("Current state: %v.", buf)
+			err = json.Unmarshal(buf, newRanch)
+			if err != nil {
+				return nil, err
+			}
+		} else if !os.IsNotExist(err) {
+			return nil, err
+		}
+	}
+
+	if err := newRanch.SyncConfig(config); err != nil {
+		return nil, err
+	}
+
+	newRanch.LogStatus()
+
+	return newRanch, nil
+}
+
+// Acquire checks out a type of resource in certain state without an owner
+// In: rtype - name of the target resource
+//     state - destination state of the resource
+//     owner - requester of the resource
+// Out: A valid Resource object on success, or
+//      ResourceNotFound error if target type resource does not exist in target state.
+func (r *Ranch) Acquire(rtype string, state string, owner string) (*common.Resource, error) {
```

Comment:
The ranch code lives in its own package. It ought to have its own unit tests to validate its behavior.

Ideally I would want more unit tests for the ranch and less for boskos.

Primarily boskos just need to check the three scenarios:
1) a resource is returned -> results in 200
2) resource is not found -> results in 404
3) sanity check the request

The ranch tests should validate the behavior of the ranch: acquiring a resource sets the owner, updates LastUpdate, matches the correct type etc. Boskos can just assume all of this is correct since it is part of the ranch logic.

Similar pattern for the other methods.